[schema registry] Avsc details leak through prototypes given to deserialized objects

Forking distinct issue from #10950 

>Decide if it is OK that deserialized objects are given a prototype from avsc causing a slight observable difference between plain data going in and data coming back.
>  - This may not be exactly what's happening here. Put together a repro and figure out exactly what causes deepEqual to fail currently.   See https://github.com/Azure/azure-sdk-for-js/pull/10890#discussion_r482449414

I debugged through this and put a repro of the underlying cause below.

Summary: `avsc` gives objects it deserializes a prototype that causes the result returned to have 7 additional inherited enumerable properties: clone, compare, isValid, toBuffer, toString, wrap, wrapped.  This is what causes chai to throw on deepEqual comparison as it considers all enumerable properties, including functions AFAICT. JSON.stringify(left) === JSON.stringify(right), which was the workaround pending investigation holds because the extra enumerable properties are functions, which JSON.stringify skips.

So now the question remains: is this OK? I am *slightly* concerned that consumers of schema-registry-avro would take a dependency on these functions, which might prevent us from replacing the avro serializer with another implementation due to the risk of breaking such uses. But maybe that is too paranoid. @xirzec Thoughts?

So far it appears to be non-trivial to strip these properties out (or make them non-enumerable if we prefer) as the return value can be a graph of objects where each of the objects would have them. I'm looking into whether the resolver arg to fromBuffer might be able to allow me to do this cleanly.

### Repro of underlying cause

``` TypeScript
import * as avro from "avsc";
import { assert } from "chai";

const schema: avro.schema.RecordType = {
  type: "record",
  name: "User",
  namespace: "com.azure.schemaregistry.samples",
  fields: [
    {
      name: "firstName",
      type: "string",
    },
    {
      name: "lastName",
      type: "string",
    },
  ],
};

const type = avro.Type.forSchema(schema);
const value = { firstName: "Nick", lastName: "Guerrera"};
const serialized = type.toBuffer(value); 
const deserialized = type.fromBuffer(serialized);

for (const key in value) {
    console.log(key);
}
console.log("");
for (const key in deserialized) {
    console.log(key);
}

assert.deepEqual(value, deserialized);
```

```
> node .\index.js
firstName
lastName

firstName
lastName
clone
compare
isValid
toBuffer
toString
wrap
wrapped

C:\Temp\rp\node_modules\chai\lib\chai\assertion.js:141
      throw new AssertionError(msg, {
      ^
AssertionError: expected { Object (firstName, lastName) } to deeply equal { Object (firstName, lastName) }
    at Object.<anonymous> (C:\Temp\rp\index.js:50:15)
    at Module._compile (internal/modules/cjs/loader.js:1075:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1096:10)
    at Module.load (internal/modules/cjs/loader.js:940:32)
    at Function.Module._load (internal/modules/cjs/loader.js:781:14)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:72:12)
    at internal/main/run_main_module.js:17:47 {
  showDiff: true,
  actual: { firstName: 'Nick', lastName: 'Guerrera' },
  expected: User { firstName: 'Nick', lastName: 'Guerrera' }
}
```


![image](https://user-images.githubusercontent.com/75470/94960779-45fa9900-04a8-11eb-99bd-004a78c57f1a.png)

--- a/sdk/schemaregistry/schema-registry-avro/package.json
+++ b/sdk/schemaregistry/schema-registry-avro/package.json
@@ -64,7 +64,7 @@
     "@azure/core-http": "^1.1.6",
     "@azure/logger": "^1.0.0",
     "@opentelemetry/api": "^0.10.2",
-    "avsc": "^5.4.21",
+    "avsc": "^5.5.1",
     "tslib": "^2.0.0",
     "buffer": "^5.2.1",
     "process": "^0.11.10"

--- a/sdk/schemaregistry/schema-registry-avro/src/schemaRegistryAvroSerializer.ts
+++ b/sdk/schemaregistry/schema-registry-avro/src/schemaRegistryAvroSerializer.ts
@@ -173,7 +173,7 @@ export class SchemaRegistryAvroSerializer {
       );
     }
 
-    const avroType = avro.Type.forSchema(JSON.parse(schemaResponse.content));
+    const avroType = this.getAvroTypeForSchema(schemaResponse.content);
     return this.cache(schemaId, schemaResponse.content, avroType);
   }
 
@@ -183,7 +183,7 @@ export class SchemaRegistryAvroSerializer {
       return cached;
     }
 
-    const avroType = avro.Type.forSchema(JSON.parse(schema));
+    const avroType = this.getAvroTypeForSchema(schema);
     if (!avroType.name) {
       throw new Error("Schema must have a name.");
     }
@@ -207,5 +207,9 @@ export class SchemaRegistryAvroSerializer {
     this.cacheByContent.set(schema, entry);
     this.cacheById.set(id, entry);
     return entry;
+  }
+
+  private getAvroTypeForSchema(schema: string) {
+    return avro.Type.forSchema(JSON.parse(schema), { omitRecordMethods: true });
   }
 }

--- a/sdk/schemaregistry/schema-registry-avro/test/schemaRegistryAvroSerializer.spec.ts
+++ b/sdk/schemaregistry/schema-registry-avro/test/schemaRegistryAvroSerializer.spec.ts
@@ -24,7 +24,7 @@ dotenv.config();
 
 chaiUse(chaiPromises);
 
-const testSchemaObject = {
+const testSchemaObject: avro.schema.RecordType = {
   type: "record",
   name: "AvroUser",
   namespace: "com.azure.schemaregistry.samples",
@@ -49,7 +49,7 @@ const testSchemaIds = [
 
 const testSchema = JSON.stringify(testSchemaObject);
 const testValue = { name: "Nick", favoriteNumber: 42 };
-const testAvroType = avro.Type.forSchema(<any>testSchemaObject);
+const testAvroType = avro.Type.forSchema(testSchemaObject, { omitRecordMethods: true });
 
 describe("SchemaRegistryAvroSerializer", function() {
   it("rejects buffers that are too small", async () => {
@@ -93,10 +93,10 @@ describe("SchemaRegistryAvroSerializer", function() {
     const schemaId = await registerTestSchema(registry);
     const serializer = await createTestSerializer(false, registry);
     const buffer = await serializer.serialize(testValue, testSchema);
-    assert.equal(0x0, buffer.readUInt32BE(0));
-    assert.equal(schemaId, buffer.toString("utf-8", 4, 36));
+    assert.strictEqual(0x0, buffer.readUInt32BE(0));
+    assert.strictEqual(schemaId, buffer.toString("utf-8", 4, 36));
     const payload = buffer.slice(36);
-    assertSameJsonRepresentation(testAvroType.fromBuffer(payload), testValue);
+    assert.deepStrictEqual(testAvroType.fromBuffer(payload), testValue);
   });
 
   it("deserializes from the expected format", async () => {
@@ -108,21 +108,21 @@ describe("SchemaRegistryAvroSerializer", function() {
 
     buffer.write(schemaId, 4, 32, "utf-8");
     payload.copy(buffer, 36);
-    assertSameJsonRepresentation(await serializer.deserialize(buffer), testValue);
+    assert.deepStrictEqual(await serializer.deserialize(buffer), testValue);
   });
 
   it("serializes and deserializes in round trip", async () => {
     let serializer = await createTestSerializer();
     let buffer = await serializer.serialize(testValue, testSchema);
-    assertSameJsonRepresentation(await serializer.deserialize(buffer), testValue);
+    assert.deepStrictEqual(await serializer.deserialize(buffer), testValue);
 
     // again for cache hit coverage on serialize
     buffer = await serializer.serialize(testValue, testSchema);
-    assertSameJsonRepresentation(await serializer.deserialize(buffer), testValue);
+    assert.deepStrictEqual(await serializer.deserialize(buffer), testValue);
 
     // throw away serializer for cache miss coverage on deserialize
     serializer = await createTestSerializer(false);
-    assertSameJsonRepresentation(await serializer.deserialize(buffer), testValue);
+    assert.deepStrictEqual(await serializer.deserialize(buffer), testValue);
 
     // thow away serializer again and cover getSchemaId instead of registerSchema
     serializer = await createTestSerializer(false);
@@ -149,17 +149,9 @@ describe("SchemaRegistryAvroSerializer", function() {
     // Deserialize buffer to value
     const deserializedValue = await serializer.deserialize(buffer);
 
-    assertSameJsonRepresentation(deserializedValue, value);
+    assert.deepStrictEqual(deserializedValue, value);
   });
 });
-
-function assertSameJsonRepresentation(actual: any, expected: any) {
-  // avsc returns objects with different prototype than our plain object literal
-  // so assert.deepEqual fails.
-  //
-  // REVIEW: Is it a problem that this avsc detail leaks?
-  assert.strictEqual(JSON.stringify(actual), JSON.stringify(expected));
-}
 
 async function createTestSerializer(
   autoRegisterSchemas = true,

